MatchTemplate with mask throws for oddly specific template image sizes / method combination

### System Information

OpenCV version 4.8.1 
OS: Windows 11
python version: 3.11.6

### Detailed description

When using the masked version of matchTemplate, having ALL of the following conditions at the same time
- providing a mask
- using method cv.TM_CCOEFF  or cv.TM_CCOEFF_NORM
- the template size is exactly (im_height - 3, im_width)

causes the following exception:
````
cv2.error: OpenCV(4.8.1) C:\Work\lib_sources\x64\opencv\modules\core\src\arithm.cpp:674: error: (-5:Bad argument) When the input arrays in add/subtract/multiply/divide functions have different types, the output array type must be explicitly specified in function 'cv::arithm_op'
````

Seems unrelated to the image /template/mask content (tested with random images, see below, but also occured in actual template matching scenario where other metrics/image size combinations work fine.)

Tested with various image/template sizes, all available methods, no mask/ with mask. See code below. Also tested with np.uint8/no.float32 dtypes (not shown).

### Steps to reproduce

````
import itertools
from collections import Counter
from enum import Enum, member
from typing import NamedTuple

import cv2 as cv
import numpy as np


class TemplateMatchModes(Enum):
    TM_SQDIFF = cv.TM_SQDIFF
    TM_SQDIFF_NORMED = cv.TM_SQDIFF_NORMED
    TM_CCORR = cv.TM_CCORR
    TM_CCORR_NORMED = cv.TM_CCORR_NORMED
    TM_CCOEFF = cv.TM_CCOEFF
    TM_CCOEFF_NORMED = cv.TM_CCOEFF_NORMED


def _random_im(shape: tuple[int, ...]) -> np.ndarray:
    return np.random.randint(low=0, high=255, size=shape, dtype=np.uint8)


def _random_mask_like(im: np.ndarray) -> np.ndarray:
    return np.random.randint(low=0, high=1, size=im.shape, dtype=np.uint8)


class MaskGenerator(Enum):
    ONES = member(np.ones_like)
    RANDOM = member(_random_mask_like)
    NONE = member(lambda x: None)


class FailureCause(NamedTuple):
    delta_size: tuple[int, int]
    mask_gen: str
    method: str


if __name__ == "__main__":

    failure_counter = Counter()
    n_image_sizes = 0
    for w, h in itertools.product(range(10, 100, 10), repeat=2):
        n_image_sizes += 1

        im_size = (h, w)
        im = _random_im(im_size)

        for dw, dh in itertools.product(range(-5, 1, 1), repeat=2):
            template_size = (h + dh, w + dw)
            template = _random_im(template_size)

            for mask_generator in MaskGenerator:
                mask = mask_generator.value(template)

                for method in TemplateMatchModes:
                    try:
                        res = cv.matchTemplate(
                            image=im,
                            templ=template,
                            method=method.value,
                            mask=mask,
                        )
                    except Exception:
                        failure_cause = FailureCause(
                            delta_size=(dh, dw), mask_gen=mask_generator.name, method=method.name
                        )
                        failure_counter[failure_cause] += 1

    print(f"{n_image_sizes=}")
    for item in failure_counter.items():
        print(item)

````

Reuslts in the following output
```
n_repeated_cases=81
(FailureCause(delta_size=(-3, 0), mask_gen='ONES', method='TM_CCOEFF'), 81)
(FailureCause(delta_size=(-3, 0), mask_gen='ONES', method='TM_CCOEFF_NORMED'), 81)
(FailureCause(delta_size=(-3, 0), mask_gen='RANDOM', method='TM_CCOEFF'), 81)
(FailureCause(delta_size=(-3, 0), mask_gen='RANDOM', method='TM_CCOEFF_NORMED'), 81)
```

### Issue submission checklist

- [X] I report the issue, it's not a question
- [X] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [ ] I updated to the latest OpenCV version and the issue is still there
- [X] There is reproducer code and related data files (videos, images, onnx, etc)

--- a/modules/imgproc/src/templmatch.cpp
+++ b/modules/imgproc/src/templmatch.cpp
@@ -850,7 +850,8 @@ static void matchTemplateMask( InputArray _img, InputArray _templ, OutputArray _
 
         // CCorr(I', T') = CCorr(I, T'*M) - sum(T'*M)/sum(M)*CCorr(I, M)
         // It does not matter what to use Mat/MatExpr, it should be evaluated to perform assign subtraction
-        Mat temp_res = img_mask_corr.mul(sum(templx_mask).div(mask_sum));
+        Mat temp_res;
+        multiply(img_mask_corr, sum(templx_mask).div(mask_sum), temp_res);
         if (img.channels() == 1)
         {
             result -= temp_res;
@@ -881,8 +882,11 @@ static void matchTemplateMask( InputArray _img, InputArray _templ, OutputArray _
             Mat img_mask2_corr(corrSize, img.type());
             crossCorr(img2, mask2, norm_imgx, Point(0,0), 0, 0);
             crossCorr(img, mask2, img_mask2_corr, Point(0,0), 0, 0);
-            temp_res = img_mask_corr.mul(Scalar(1.0, 1.0, 1.0, 1.0).div(mask_sum))
-                           .mul(img_mask_corr.mul(mask2_sum.div(mask_sum)) - 2 * img_mask2_corr);
+            Mat temp_res1;
+            multiply(img_mask_corr, Scalar(1.0, 1.0, 1.0, 1.0).div(mask_sum), temp_res1);
+            Mat temp_res2;
+            multiply(img_mask_corr, mask2_sum.div(mask_sum), temp_res2);
+            temp_res = temp_res1.mul(temp_res2 - 2 * img_mask2_corr);
             if (img.channels() == 1)
             {
                 norm_imgx += temp_res;

--- a/modules/imgproc/test/test_templmatchmask.cpp
+++ b/modules/imgproc/test/test_templmatchmask.cpp
@@ -275,4 +275,16 @@ INSTANTIATE_TEST_CASE_P(MultiChannelMask, Imgproc_MatchTemplateWithMask2,
         Values(cv::TM_SQDIFF, cv::TM_SQDIFF_NORMED, cv::TM_CCORR, cv::TM_CCORR_NORMED,
                cv::TM_CCOEFF, cv::TM_CCOEFF_NORMED)));
 
+TEST(Imgproc_MatchTemplateWithMask, bug_26389) {
+    const Mat image = Mat::ones(Size(10, 10), CV_8UC1);
+    const Mat templ = Mat::ones(Size(10, 7), CV_8UC1);
+    const Mat mask = Mat::ones(Size(10, 7), CV_8UC1);
+
+    for (const int method : {TM_CCOEFF, TM_CCOEFF_NORMED})
+    {
+        Mat result;
+        matchTemplate(image, templ, result, method, mask);
+    }
+}
+
 }} // namespace

